python/cudf/cudf/tests/test_orc.py

```diff
@@ -1839,3 +1839,34 @@ def test_orc_writer_cols_as_map_type_error():
         TypeError, match="cols_as_map_type must be a list of column names."
     ):
         df.to_orc(buffer, cols_as_map_type=1)
+
+
+@pytest.mark.parametrize("engine", ["cudf", "pyarrow"])
+def test_negative_timestamp(tmpdir, engine):
+    expected = cudf.DataFrame(
+        {
+            "a": [
+                pd.Timestamp("1969-12-31 23:59:58.000999"),
+                pd.Timestamp("1969-12-31 23:59:58.001001"),
+                pd.Timestamp("1839-12-24 03:58:56.000826"),
+            ]
+        }
+    )
+    cudf_fname = tmpdir.join("cudf_neg_ts.orc")
+    expected.to_orc(cudf_fname)
+
+    cudf_got = cudf.read_orc(cudf_fname, engine=engine)
+    assert_eq(expected, cudf_got)
+    assert_eq(expected, pd.read_orc(cudf_fname))
+    assert pyarrow.orc.ORCFile(cudf_fname).read().equals(cudf_got.to_arrow())
+
+    pyorc_fname = tmpdir.join("pyorc_neg_ts.orc")
+    pyorc_table = pa.Table.from_pandas(
+        expected.to_pandas(), preserve_index=False
+    )
+    pyarrow.orc.write_table(pyorc_table, pyorc_fname)
+
+    cudf_got = cudf.read_orc(pyorc_fname, engine=engine)
+    assert_eq(expected, cudf_got)
+    assert_eq(expected, pd.read_orc(pyorc_fname))
+    assert pyarrow.orc.ORCFile(pyorc_fname).read().equals(cudf_got.to_arrow())
```

Comment:
Need to use `cudf_got` here as comparison with `expected` fails, even though `comparison` between `expected` and `cudf_got` passes.

Reply:
I'm trying to understand this. We do `assert_eq(expected, cudf_got)`, so clearly there's something that is special about `cudf_got` that makes it align with the pyarrow ORC reader that `expected` doesn't have. Why does comparison to `expected.to_arrow()` fail?

Reply:
I think this deserves a comment in the test, because it looks like it breaks the expectation that equality is transitive.

Reply:
```
E       AssertionError: assert False
E        +  where False = <built-in method equals of pyarrow.lib.Table object at 0x7fdd72ffcb70>(pyarrow.Table\na: timestamp[us]\n----\na: [[1969-12-31 23:59:59.000123,1969-12-31 23:59:58.000999,1969-12-31 23:59:58.001001,1839-12-24 03:58:56.000826]])
E        +    where <built-in method equals of pyarrow.lib.Table object at 0x7fdd72ffcb70> = pyarrow.Table\na: timestamp[ns]\n----\na: [[1969-12-31 23:59:59.000123000,1969-12-31 23:59:58.000999000,1969-12-31 23:59:58.001001000,1839-12-24 03:58:56.000826000]].equals
E        +      where pyarrow.Table\na: timestamp[ns]\n----\na: [[1969-12-31 23:59:59.000123000,1969-12-31 23:59:58.000999000,1969-12-31 23:59:58.001001000,1839-12-24 03:58:56.000826000]] = <bound method ORCFile.read of <pyarrow.orc.ORCFile object at 0x7fdd6b5c5b80>>()
E        +        where <bound method ORCFile.read of <pyarrow.orc.ORCFile object at 0x7fdd6b5c5b80>> = <pyarrow.orc.ORCFile object at 0x7fdd6b5c5b80>.read
E        +          where <pyarrow.orc.ORCFile object at 0x7fdd6b5c5b80> = <class 'pyarrow.orc.ORCFile'>(local('/tmp/pytest-of-vukasin/pytest-1/test_negative_timestamp_pyarro0/cudf_neg_ts.orc'))
E        +            where <class 'pyarrow.orc.ORCFile'> = <module 'pyarrow.orc' from '/home/vukasin/anaconda3/envs/cudf_dev/lib/python3.9/site-packages/pyarrow/orc.py'>.ORCFile
E        +              where <module 'pyarrow.orc' from '/home/vukasin/anaconda3/envs/cudf_dev/lib/python3.9/site-packages/pyarrow/orc.py'> = pyarrow.orc
E        +    and   pyarrow.Table\na: timestamp[us]\n----\na: [[1969-12-31 23:59:59.000123,1969-12-31 23:59:58.000999,1969-12-31 23:59:58.001001,1839-12-24 03:58:56.000826]] = <bound method DataFrame.to_arrow of                            a\n0 1969-12-31 23:59:59.000123\n1 1969-12-31 23:59:58.000999\n2 1969-12-31 23:59:58.001001\n3 1839-12-24 03:58:56.000826>()
E        +      where <bound method DataFrame.to_arrow of                            a\n0 1969-12-31 23:59:59.000123\n1 1969-12-31 23:59:58.000999\n2 1969-12-31 23:59:58.001001\n3 1839-12-24 03:58:56.000826> =                            a\n0 1969-12-31 23:59:59.000123\n1 1969-12-31 23:59:58.000999\n2 1969-12-31 23:59:58.001001\n3 1839-12-24 03:58:56.000826.to_arrow
```

Reply:
I'm having trouble seeing why the comparison failed from your code snippet. The values look the same. Is there differing metadata or a different index or something that I'm not seeing? Can we debug how the `equals` method works to see what is mismatched?

Feel free to give up if this gets too hairy -- just leave a comment in the code instead of on this PR:
```suggestion
    # Need to use cudf_got here as comparison with expected fails, even though
    # comparison between expected and cudf_got passes.
    assert pyarrow.orc.ORCFile(pyorc_fname).read().equals(cudf_got.to_arrow())
```

Reply:
Modified the test to use `assert_eq` here instead and the comparison works as expected.